source/val/Construct.cpp

```diff
@@ -26,19 +26,51 @@
 
 #include "val/Construct.h"
 
+#include <cassert>
+#include <cstddef>
+
 namespace libspirv {
 
-Construct::Construct(BasicBlock* header_block, BasicBlock* merge_block,
-                     BasicBlock* continue_block)
-    : header_block_(header_block),
-      merge_block_(merge_block),
-      continue_block_(continue_block) {}
+Construct::Construct(ConstructType type, BasicBlock* dominator,
+                     BasicBlock* exit, std::vector<Construct*> constructs)
+    : type_(type),
+      corresponding_constructs_(constructs),
+      base_dominator_(dominator),
+      exit_block_(exit) {}
+
+ConstructType Construct::get_type() const { return type_; }
+
+const std::vector<Construct*>& Construct::get_corresponding_constructs() const {
+  return corresponding_constructs_;
+}
+std::vector<Construct*>& Construct::get_corresponding_constructs() {
+  return corresponding_constructs_;
+}
+
+bool ValidateConstructSize(ConstructType type, size_t size) {
+  switch (type) {
+    case ConstructType::kSelection: return size == 0;
+    case ConstructType::kContinue:  return size == 1;
```

Comment:
I am creating constructs for kLoop and kContinue even if they represent the same block. So they will always have one corresponding construct.

But the kCase ... case is incorrect. It should be size >= 1;